uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js

```diff
@@ -48,7 +48,7 @@ define([
     vm.contactName = null; // contact name of the owner of leave request
     vm.errors = [];
     vm.fileUploader = null;
-    vm.staffMemberSelectionComplete = false;
+    vm.haveCommentsBeenUpdated = false;
```

Comment:
This is not ideal, because the request controller should not know anything about the properties of the child controllers(comments or files). But we already had `fileUploader` object and now you have added `haveCommentsBeenUpdated`. I dont like having different ad hoc variables like this. So I have this plan

We already have a `checkSubmitConditions` function in `request.controller` which is used to know if request can be submitted for the details tab, similar `fileUploader` and `haveCommentsBeenUpdated` are present for other tabs. Lets create 3 functions `checkSubmitConditionsForDetailsTab`, `checkSubmitConditionsForCommentsTab` and `checkSubmitConditionsForFilesTab` and those functions will have their internal logic. 

And in request controller we can update the `canSubmit` as
```javascript
function canSubmit () {
      var canSubmit = vm.checkSubmitConditionsForDetailsTab() && vm.checkSubmitConditionsForCommentsTab() && vm.checkSubmitConditionsForFilesTab()

      // check if user has changed any attribute or added a comment
      if (vm.isMode('edit')) {
        canSubmit = canSubmit && hasRequestChanged();
      }

      // check if manager has changed status
      if (vm.canManage && vm.requestStatuses) {
        // awaiting_approval will not be available in vm.requestStatuses if manager has changed selection
        canSubmit = canSubmit && !!vm.getStatusFromValue(vm.newStatusOnSave);
      }

      // check if the selected date period is in absence period
      canSubmit = canSubmit && !!vm.period.id;

      return canSubmit && !vm.isMode('view');
    }
```
and `hasRequestChanged` as
```javascript
function hasRequestChanged () {
      // using angular.equals to automatically ignore the $$hashkey property
      return !angular.equals(
        initialLeaveRequestAttributes,
        vm.request.attributes()
      ) || (vm.canManage && vm.newStatusOnSave);
    }
```